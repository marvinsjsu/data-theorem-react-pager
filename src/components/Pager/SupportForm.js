import React from 'react';
import PropTypes from 'prop-types';

export default class SupportForm extends React.Component {

  static propTypes = {
    sendMessage: PropTypes.func.isRequired,
    supportRequestUrl: PropTypes.string.isRequired
  };

  state = {
    name: '',
    email: '',
    message: ''
  };

  onSubmit = (e) => {
    e.preventDefault();
    console.log('onSubmit');
    const { sendMessage } = this.props;
    sendMessage({...this.state});
  };

  catchEnter = (e) => {

    if (e.key === 'Enter') {
      e.preventDefault();
    }
  };

  onChange = (key, e) => {
    this.setState({
      [key]: e.target.value
    });
  };

  render () {
    const { name, email, message } = this.state;
    const { closeSupportDialog } = this.props;

    return (
      <div className='container--form' id='support-form'>
        <form onSubmit={this.onSubmit}>
          <h4 className='title'>Support Request Form</h4>
          <button
            className="container--form__close"
            onClick={closeSupportDialog}
          >
            &times;
          </button>
          <div className='row'>
            <input
              id='name'
              className='input'
              type='text'
              name='name'
              value={name}
              placeholder='name'
              onChange={(e) => this.onChange('name', e)}
              onKeyDown={this.catchEnter}
              required
            />
          </div>

          <div className='row'>
            <input
              id='email'
              className='input'
              type='email'
              name='email'
              value={email}
              placeholder='email'
              onChange={(e) => this.onChange('email', e)}
              onKeyDown={this.catchEnter}
              required
            />
          </div>

          <div className='row'>
            <textarea
              id='message'
              className='textarea--message'
              name='message'
              value={message}
              placeholder='message'
              onChange={(e) => this.onChange('message', e)}
              onKeyDown={this.catchEnter}
              required
            />
          </div>

          <div className='row'>
            <button
              type='submit'
              className='btn'
              disabled={!name || !email || !message}
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }
};
