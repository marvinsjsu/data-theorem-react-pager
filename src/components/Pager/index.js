import React from 'react';
import PropTypes from 'prop-types';

import { getPage } from '../../utils/api';

const GO_BACK = 1;
const GO_PREV = -1;

export default class Pager extends React.Component {

  static propTypes = {
    pages: PropTypes.array.isRequired,
    getLabel: PropTypes.func.isRequired,
    children: PropTypes.func.isRequired,
    supportRequestUrl: PropTypes.string,
    pageInfoUrl: PropTypes.func
  };

  static defaultProps = {
    supportRequestUrl: null,
    pageInfoUrl: null
  };

  state = {
    initialized: false,
    page: this.props.pages[0],
    pages: this.props.pages,
    pageIndex: 0,
    pageInfo: {},
    pageInfoIsLoading: true,
    pageInfoError: null
  };

  componentDidMount () {
    const { pages, children, getLabel } = this.props;

    if (!children || !pages || !getLabel) return false;
    this.mounted = true;
    this._loadPageUrl();
  };

  componentWillUnmount () {
    this.mounted = false;
  };

  currentPageLabel = () => {
    const { pageIndex } = this.state;
    const { getLabel } = this.props;
    return getLabel(pageIndex);
  };

  pageLabels = () => {
    const { pages } = this.state;
    const { getLabel } = this.props;

    return pages.map((page, idx) => (getLabel(idx)));
  };

  goPrevious = () => {
    this._movePage(GO_PREV);
  };

  goNext = () => {
    this._movePage(GO_BACK);
  };

  goToLabel = (label) => {
    const { pages } = this.state;
    const pageIndex = this.pageLabels().indexOf(label);
    const page = pages[pageIndex];

    if (pageIndex !== -1) {
      this.setState({
        pageIndex,
        page
      });
    }
  };

  openSupportDialog = () => {

  };

  _getPageIndex = (step, currIndex, pageCount) => {
    let newPageIndex = currIndex + step;
    let mod = newPageIndex % pageCount;
    return mod > 0 ? mod : Math.abs(pageCount + mod) % pageCount;
  };

  _movePage = (step) => {
    this.setState((currState) => {
      const newPageIndex = this._getPageIndex(step, currState.pageIndex, currState.pages.length);
      const newPage = currState.pages[newPageIndex];
      return {
        pageIndex: newPageIndex,
        page: newPage
      }
    }, this._loadPageUrl);
  };

  _loadPageUrl = () => {
    const { pageInfoUrl } = this.props;
    const pageLabel = this.currentPageLabel();

    if (pageInfoUrl && pageLabel) {
      getPage(pageInfoUrl(pageLabel))
        .then((pageInfo) => {


console.log('pageInfo: ', pageInfo);

          if (this.mounted && pageInfo) {
            this.setState({
              pageInfo,
              pageInfoIsLoading: true
            });
          }
        })
        .catch((error) => {


console.log('error: ', error);


          if (this.mounted) {
            this.setState({
              pageInfoError: error
            });
          }
        });
    }
  };

  render () {
    const { children: Component } = this.props;

    return (
      <Component
        {...this.state}
        goNext={this.goNext}
        goPrevious={this.goPrevious}
        goToLabel={this.goToLabel}
        openSupportDialog={this.openSupportDialog}
        pageLabels={this.pageLabels()}
        currentPageLabel={this.currentPageLabel()}
      />
    );
  }
}