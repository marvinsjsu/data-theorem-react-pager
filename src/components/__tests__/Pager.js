import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Pager from '../Pager';
import SupportForm from '../Pager/SupportForm';

jest.mock('../../utils/api');
Enzyme.configure({ adapter: new Adapter() });

describe(`The Pager component`, () => {

  it (`renders as expected`, () => {
    const component = renderer.create(pageComponent);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test (`pageIndex initial value and after goNext/goPrevious`, () => {
    const wrapper = mount(pageComponent);
    expect(wrapper.state().pageIndex).toBe(0);

    const { goNext, goPrevious } = wrapper.children().props();

    goNext();
    expect(wrapper.state().pageIndex).toBe(1);

    goPrevious();
    expect(wrapper.state().pageIndex).toBe(0);

    goPrevious();
    expect(wrapper.state().pageIndex).toBe(employees.length - 1);

    goNext();
    expect(wrapper.state().pageIndex).toBe(0);
  });

  test (`update showSupportDialog when openSupportDialog is invoked`, () => {
    const wrapper = mount(pageComponent);
    expect(wrapper.state().pageIndex).toBe(0);

    const { openSupportDialog, closeSupportDialog } = wrapper.children().props();

    openSupportDialog();
    expect(wrapper.state().showSupportDialog).toBe(true);
  });
});


const employees = [
  {
    last_name: 'mante',
    first_name: 'marvin',
    department: 'engineering',
    salary: 'underpaid'
  },
  {
    last_name: 'oani',
    first_name: 'aimee',
    department: 'teaching',
    salary: 'definitely underpaid',
  },
  {
    last_name: 'mante',
    first_name: 'sophia',
    department: 'exploring',
    salary: 'more than we make combined'
  },
];

const pageComponent = (
  <Pager
    pages={employees.map((employee, idx) => (
      <ul key={idx}>
        <li>Name: {employee.last_name}, {employee.first_name}</li>
        <li>Department: {employee.department}</li>
        <li>Salary: ${employee.salary}</li>
      </ul>
    ))}
    getLabel={
      i => `${employees[i].last_name}, ${employees[i].first_name}`
    }
    pageInfoUrl={(label) => `https://www.example.com/employees/info?label=${label}`}
    supportRequestUrl="https://www.example.com/support"
  >
    {({
      page,
      goPrevious,
      goNext,
      goToLabel,
      currentPageLabel,
      pageLabels,
      openSupportDialog,
      pageInfoIsLoading,
      pageInfoError,
      pageInfo,
    })=>(
      <React.Fragment>
        <div>
          <select onChange={e => goToLabel(e.target.value)} defaultValue={currentPageLabel}>
            {pageLabels.map(label => (
              <option
                key={label}
                value={label}
              >
                {label}
              </option>
            ))}
          </select>
          <button onClick={goPrevious}>Previous</button>
          <button onClick={goNext}>Next</button>
          <button onClick={openSupportDialog}>Help</button>
        </div>
        <div>
          {page}
        </div>
        {pageInfoIsLoading && (
          <div>Loading more info...</div>
        )}
        {pageInfoError && (
          <div>Error fetching info: {pageInfoError}</div>
        )}
        {pageInfo && (
          <div>
            # of Likes: {pageInfo.likes}
          </div>
        )}
      </React.Fragment>
    )}
  </Pager>
);