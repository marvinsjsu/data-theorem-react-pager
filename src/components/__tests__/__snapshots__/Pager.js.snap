// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Pager component renders as expected 1`] = `
Array [
  <div>
    <select
      defaultValue="mante, marvin"
      onChange={[Function]}
    >
      <option
        value="mante, marvin"
      >
        mante, marvin
      </option>
      <option
        value="oani, aimee"
      >
        oani, aimee
      </option>
      <option
        value="mante, sophia"
      >
        mante, sophia
      </option>
    </select>
    <button
      onClick={[Function]}
    >
      Previous
    </button>
    <button
      onClick={[Function]}
    >
      Next
    </button>
    <button
      onClick={[Function]}
    >
      Help
    </button>
  </div>,
  <div>
    <ul>
      <li>
        Name: 
        mante
        , 
        marvin
      </li>
      <li>
        Department: 
        engineering
      </li>
      <li>
        Salary: $
        underpaid
      </li>
    </ul>
  </div>,
  <div>
    Loading more info...
  </div>,
]
`;
